lib/DiscoveryStreamFeed.jsm

```diff
@@ -0,0 +1,148 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+ChromeUtils.import("resource://gre/modules/XPCOMUtils.jsm");
+XPCOMUtils.defineLazyGlobalGetters(this, ["fetch"]);
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+
+const {actionTypes: at, actionCreators: ac} = ChromeUtils.import("resource://activity-stream/common/Actions.jsm", {});
+const {PersistentCache} = ChromeUtils.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+
+const CACHE_KEY = "discovery_stream";
+const LAYOUT_UPDATE_TIME = 30 * 60 * 1000; // 30 minutes
+const CONFIG_PREF_NAME = "browser.newtabpage.activity-stream.discoverystream.config";
+
+this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
+  constructor() {
+    // Internal state for checking if we've intialized all our data
+    this.loaded = false;
+
+    // Persistent cache for remote endpoint data.
+    this.cache = new PersistentCache(CACHE_KEY, true);
+    // Internal in-memory cache for parsing json prefs.
+    this._prefCache = {};
+  }
+
+  get config() {
+    if (this._prefCache.config) {
+      return this._prefCache.config;
+    }
+    try {
+      this._prefCache.config = JSON.parse(Services.prefs.getStringPref(CONFIG_PREF_NAME, ""));
+    } catch (e) {
+      // istanbul ignore next
+      Cu.reportError(`Could not parse preference. Try resetting ${CONFIG_PREF_NAME} in about:config.`);
```

Comment:
Should this at least set `_prefCache.config = {}`? Otherwise `.config` could be undefined and result in brokeness elsewhere `.config.enabled`